server/src/main/scala/org/http4s/server/middleware/Throttle.scala

```diff
@@ -0,0 +1,90 @@
+package org.http4s.server.middleware
+
+import org.http4s.{Http, Response, Status}
+import cats.data.Kleisli
+import cats.effect.{Concurrent, Timer}
+import cats.effect.concurrent.Ref
+import scala.concurrent.duration._
+import cats.implicits._
+import fs2.Stream
+
+sealed trait TokenAvailability
```

Comment:
`Token` makes my mind think about authentication / Authorization. Is this a common terminology used for throttling as well?

Reply:
It is. The algorithm Keir used here is commonly known as a "Token bucket". 
If you think about it _is_ similar to authentication, in both cases a token grants you the capability to do something.

Reply:
@SystemFw fair, thanks for the clarification :)